nixos/modules/services/cluster/kubernetes/default.nix

```diff
@@ -265,7 +266,7 @@ in {
           to members of the cluster. This address must be reachable by the rest
           of the cluster.
         '';
-        default = null;
+        default = cfg.apiserver.bindAddress;
```

Comment:
I wonder what happens if "0.0.0.0" is advertised. Does that make sense to be the default?

Reply:
I thought about that myself, however I think it is the correct way to go about it, since this is what the apiserver falls back to itself, according to the docs: https://kubernetes.io/docs/reference/command-line-tools-reference/kube-apiserver/ . 

```
--advertise-address ip

   The IP address on which to advertise the apiserver to members of the cluster. 
   This address must be reachable by the rest of the cluster.
   If blank, the --bind-address will be used. 
   If --bind-address is unspecified, the host's default interface will be used.
```

... but perhaps it's better to let the apiserver itself to the fallback through 1. bind-address, 2. host-address. Thus making it default = null and implementing with `optionalString cfg.advertiseAddress --advertise-address cfg.advertiseAddress`. Then however, we would need to **not** use cfg.advertiseAddress in the kubelet config.

.. a third option is therefore, don't set any default value for the advertiseAddress and require it to be configured? This is the best alternative I can think of.

Of course all of this boils down to: How plug'n'play do we want the module to be?

Reply:
Since advertising a service at 0.0.0.0 is practically impossible I was curious about what IP is advertised to cluster members if this is specified or as your quote says if fallen back to.

Someone had already done the code digging: https://github.com/kubernetes/kubernetes/issues/58686#issuecomment-360677218

* [If not set or unspecified it will default to the secure external interface](https://github.com/kubernetes/apiserver/blob/d955b8a826dde3566e9dc4a51493b4afc1b51a7b/pkg/server/options/server_run_options.go#L77-L78)
* [0.0.0.0 is considered unspecified](https://golang.org/src/net/ip.go?h=IsUnspecified#L113)

I don't know if we can easily reproduce `secure.DefaultExternalAddress()` which is part of the fallback mechanism. To me it seems sensible to leave the default at `null` and thereby let the kube-apiserver figure it out, while providing the option to set it if needed. Do you see any difficulty with that option?

Reply:
Of course someone already digged into that! :+1: :)

I'm ok with that solution, only problem I see is this line, which uses the advertiseAddress: https://github.com/johanot/nixpkgs/blob/k8s-improvements/nixos/modules/services/cluster/kubernetes/default.nix#L76

If we let kube-apiserver decide advertiseAddress at runtime, we cannot sanely set a default value of the kubeconfig server option either ? Can we ?

Reply:
Would `if cfg.apiserver.advertiseAddress != null then cfg.apiserver.advertiseAddress else "127.0.0.1`
work? It seems to be a sane assumption that the kubeconfig is generated for the clusternode itself. If the user needs anything different then she needs to fill in address information manually anyway.

Reply:
mkKubeConfigOptions is used for all kubelets in the cluster. For any non-master node in a multi-node cluster, 127.0.0.1 will not work. I think however that it's good enough to set as default value. At least it will work for a singlenode cluster and for components co-located with the apiserver. Will add the change to the PR now.

Reply:
Reminds me that, we better put some work into deprecating those insecure ports... Support for them might be completely gone in v1.11.. Well... Next PR.